deployment.go

```diff
@@ -82,21 +85,45 @@ func (c *DeploymentControl) Delete(ctx context.Context, cascade bool) error {
 	c.Infof("delete %v", formatMeta(c.deployment.ObjectMeta))
 
 	deployments := c.Client.Extensions().Deployments(c.deployment.Namespace)
-	currentDeployment, err := deployments.Get(c.deployment.Name)
+	currentDeployment, err := deployments.Get(c.deployment.Name, metav1.GetOptions{})
 	if err != nil {
 		return ConvertError(err)
 	}
+
+	pods := c.Client.Core().Pods(c.deployment.Namespace)
+	currentPods, err := c.collectPods(currentDeployment)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
 	if cascade {
-		// scale deployment down to delete all replicas and pods
+		// scale deployment down to delete the pods
 		var replicas int32
 		currentDeployment.Spec.Replicas = &replicas
 		currentDeployment, err = deployments.Update(currentDeployment)
 		if err != nil {
 			return ConvertError(err)
 		}
 	}
-	err = deployments.Delete(c.deployment.Name, nil)
-	return ConvertError(err)
+	deletePolicy := metav1.DeletePropagationForeground
+	err = deployments.Delete(c.deployment.Name, &metav1.DeleteOptions{
+		PropagationPolicy: &deletePolicy,
+	})
+	if err != nil {
+		return ConvertError(err)
+	}
+
+	errDelete := waitForObjectDeletion(func() error {
+		_, err := deployments.Get(c.deployment.Name, metav1.GetOptions{})
+		return ConvertError(err)
+	})
+
+	// wait until all Pods have been cleaned up
+	err = waitForPods(pods, currentPods, *c.Entry)
+	if err == nil {
+		err = errDelete
```

Comment:
Can you explain this logic? Why do we not return errDelete above and assign it here instead?

Reply:
you're right - we could return an error sooner.